CMakeLists.txt

```diff
@@ -163,12 +163,12 @@ if(CMAKE_COMPILER_IS_GNUCXX OR MINGW OR
     # set Advanced Vector Extensions (AVX)
     if(USE_AVX AND COMPILER_HAS_AVX_FLAG)
     	add_definitions(-DCNN_USE_AVX)
-        set(EXTRA_C_FLAGS "${EXTRA_C_FLAGS} -mavx")
+        set(EXTRA_C_FLAGS "${EXTRA_C_FLAGS} -mavx -mfma")
```

Comment:
@nyanp 

FMA3 instruction set are not supported by Intel Sandy Bridge cpus and Ivy Bridge cpus (though I read that it's supported by some of AMD nondominant processors...), so always coupling -mavx compiler flag with -mfma compiler flag may cause troubles.

I still haven't used the intrinsics for FMA3 instruction set in AVX backend code at ( https://github.com/nyanp/tiny-cnn/blob/feat/generic-computational-graph/tiny_cnn/core/kernels/avx_kernel_common.h ) yet. In the AVX backend, I've tried to use only AVX intrinsic functions and avoided AVX2 and FMA3 intrinsic functions. 

I thought I could achieve noticeable performance gains with AVX2/FMA3 intrinsinc functions but I have failed so far.

It would be better if the backend checks the availabilities of instructions with cpuid on x86/x64 and can automatically select the most performant routines. Maybe in the distant future.